File: frontend/lib/src/models/audio.dart
import 'package:json_annotation/json_annotation.dart';
import '../config/env_config.dart';
import 'package:flutter/foundation.dart';

part 'audio.g.dart';

@JsonSerializable()
class UserInfo {
  final String id;
  final String username;
  @JsonKey(name: 'avatar_url')
  final String? avatarUrl;

  UserInfo({
    required this.id,
    required this.username,
    this.avatarUrl,
  });

  factory UserInfo.fromJson(Map<String, dynamic> json) => _$UserInfoFromJson(json);
  Map<String, dynamic> toJson() => _$UserInfoToJson(this);
}

@JsonSerializable()
class Track {
  final String id;
  final String title;
  final String? description;
  @JsonKey(name: 'audio_url')
  final String audio_url;
  final String artist;
  final double duration;
  @JsonKey(name: 'cover_url')
  final String? cover_url;
  @JsonKey(name: 'gif_url')
  final String? gif_url;
  @JsonKey(name: 'user_id')
  final String userId;
  @JsonKey(name: 'username')
  final String? username;
  @JsonKey(name: 'user_avatar')
  final String? userAvatar;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;

  // Helper methods for full URLs
  String get fullAudioUrl => audio_url.startsWith('http') 
    ? audio_url 
    : '${EnvConfig.apiBaseUrl}/audio/stream/$id';
    
  String? get fullCoverUrl => cover_url == null 
    ? null 
    : cover_url!.startsWith('http') 
      ? cover_url 
      : '${EnvConfig.staticBaseUrl}${cover_url!.startsWith('/') ? cover_url : '/static/images/$cover_url'}';

  String? get fullUserAvatar => userAvatar == null 
    ? null 
    : userAvatar!.startsWith('http') 
      ? userAvatar 
      : '${EnvConfig.staticBaseUrl}${userAvatar!.startsWith('/') ? userAvatar : '/static/images/$userAvatar'}';

  String? get fullGifUrl {
    if (gif_url == null) return null;
    debugPrint('Original gif URL: $gif_url');
    debugPrint('Static base URL: ${EnvConfig.staticBaseUrl}');
    
    // Always treat GIF URLs as relative paths under /static/gif/
    final url = gif_url!.startsWith('/static/gif/') 
        ? '${EnvConfig.staticBaseUrl}$gif_url'
        : '${EnvConfig.staticBaseUrl}/static/gif/${gif_url!.split('/').last}';
    
    debugPrint('Full gif URL: $url');
    return url;
  }

  Track({
    required this.id,
    required this.title,
    this.description,
    required this.audio_url,
    required this.artist,
    required this.duration,
    this.cover_url,
    this.gif_url,
    required this.userId,
    this.username,
    this.userAvatar,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Track.fromJson(Map<String, dynamic> json) {
    // Convert integer IDs to strings if needed
    final modifiedJson = Map<String, dynamic>.from(json);
    if (json['id'] is int) {
      modifiedJson['id'] = json['id'].toString();
    }
    if (json['user_id'] is int) {
      modifiedJson['user_id'] = json['user_id'].toString();
    }
    return _$TrackFromJson(modifiedJson);
  }
  Map<String, dynamic> toJson() => _$TrackToJson(this);
}


@JsonSerializable()
class RecentlyPlayed {
  final String id;
  final Track track;
  @JsonKey(name: 'track_id')
  final String trackId;
  @JsonKey(name: 'played_at')
  final DateTime playedAt;
  @JsonKey(name: 'user_id', required: false)
  final String? userId;

  RecentlyPlayed({
    required this.id,
    required this.track,
    required this.trackId,
    required this.playedAt,
    this.userId,
  });

  factory RecentlyPlayed.fromJson(Map<String, dynamic> json) {
    // Convert integer IDs to strings if needed
    final modifiedJson = Map<String, dynamic>.from(json);
    if (json['id'] is int) {
      modifiedJson['id'] = json['id'].toString();
    }
    if (json['user_id'] is int) {
      modifiedJson['user_id'] = json['user_id'].toString();
    }
    if (json['track_id'] is int) {
      modifiedJson['track_id'] = json['track_id'].toString();
    }
    return _$RecentlyPlayedFromJson(modifiedJson);
  }
  Map<String, dynamic> toJson() => _$RecentlyPlayedToJson(this);

  // Delegate properties to track for convenience
  String get title => track.title;
  String get description => track.description ?? '';
  String get audio_url => track.audio_url;
  String get artist => track.artist;
  double get duration => track.duration;
  String? get cover_url => track.cover_url;
}

@JsonSerializable()
class PlaylistCreate {
  final String name;
  final String description;

  PlaylistCreate({
    required this.name,
    required this.description,
  });

  factory PlaylistCreate.fromJson(Map<String, dynamic> json) =>
      _$PlaylistCreateFromJson(json);
  Map<String, dynamic> toJson() => _$PlaylistCreateToJson(this);
}

@JsonSerializable()
class PlaylistUpdate {
  final String? name;
  final String? description;

  PlaylistUpdate({
    this.name,
    this.description,
  });

  factory PlaylistUpdate.fromJson(Map<String, dynamic> json) =>
      _$PlaylistUpdateFromJson(json);
  Map<String, dynamic> toJson() => _$PlaylistUpdateToJson(this);
}

File: backend/app/schemas/audio.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime

class TrackBase(BaseModel):
    title: str
    description: Optional[str] = None
    artist: str
    duration: float
    audio_url: str
    cover_url: Optional[str] = None
    gif_url: Optional[str] = '/static/gif/kafka_night.gif'
    user_id: str

class UserInfo(BaseModel):
    id: str
    username: str
    avatar_url: Optional[str] = None

class TrackCreate(TrackBase):
    pass

class Track(TrackBase):
    id: str
    created_at: datetime
    updated_at: datetime
    username: Optional[str] = None
    user_avatar: Optional[str] = None

    @staticmethod
    def from_orm(db_obj):
        # Get the user info
        username = db_obj.user.username if db_obj.user else None
        user_avatar = db_obj.user.avatar_url if db_obj.user else None
        
        # Create a dict of the database object
        obj_dict = {
            "id": db_obj.id,
            "title": db_obj.title,
            "description": db_obj.description,
            "artist": db_obj.artist,
            "duration": db_obj.duration,
            "audio_url": db_obj.audio_url,
            "cover_url": db_obj.cover_url,
            "gif_url": db_obj.gif_url,
            "user_id": db_obj.user_id,
            "username": username,
            "user_avatar": user_avatar,
            "created_at": db_obj.created_at,
            "updated_at": db_obj.updated_at,
        }
        return Track(**obj_dict)

    class Config:
        orm_mode = True

class PlaylistBase(BaseModel):
    name: str
    description: Optional[str] = None
    cover_url: Optional[str] = None

class PlaylistCreate(PlaylistBase):
    pass

class PlaylistUpdate(PlaylistBase):
    pass

class Playlist(PlaylistBase):
    id: str
    user_id: str
    creator_name: str
    created_at: datetime
    updated_at: datetime
    track_count: int
    tracks: Optional[List[Track]] = None

    @staticmethod
    def from_orm(db_obj):
        # Get the creator name from the user relationship
        creator_name = db_obj.user.username if db_obj.user else "System"
        
        # Create a dict of the database object
        obj_dict = {
            "id": db_obj.id,
            "name": db_obj.name,
            "description": db_obj.description,
            "user_id": db_obj.user_id,
            "creator_name": creator_name,
            "cover_url": db_obj.cover_url,
            "tracks": db_obj.tracks,
            "track_count": len(db_obj.tracks),
            "created_at": db_obj.created_at,
            "updated_at": db_obj.updated_at,
        }
        return Playlist(**obj_dict)

    class Config:
        orm_mode = True

class PlaylistAddTrack(BaseModel):
    track_id: str

class PlaylistRemoveTrack(BaseModel):
    track_id: str

class RecentlyPlayedCreate(BaseModel):
    track_id: str

class RecentlyPlayed(BaseModel):
    id: str
    user_id: str
    track_id: str
    played_at: datetime
    track: Track

    class Config:
        orm_mode = True

class AudioResponse(BaseModel):
    success: bool
    message: str
    data: Optional[dict] = None

    class Config:
        orm_mode = True
